File: mygreeterv3csharp/server/Src/Server/Handler/ListResourceGroups.cs
using AKSMiddleware;
using Google.Protobuf.WellKnownTypes;
using Grpc.Core;
using MyGreeterCsharp.Api.V1;
using Serilog;

namespace MyGreeterCsharp.Server.Handler;

public partial class MyGreeterCsharpServer
{
    public override async Task<ListResourceGroupResponse> ListResourceGroups(Empty request, ServerCallContext context)
    {
        if (_resourceGroups == null)
        {
            Log.Logger.WithCtx(context).Warning("ResourceGroupClient is nil in ListResourceGroups(), azuresdk feature is likely disabled");
            throw new RpcException(new Status(StatusCode.Unavailable, "ResourceGroupClient is nil in ListResourceGroups(), azuresdk feature is likely disabled"));
        }

        var resourceGroupList = new List<ResourceGroup>();

        try
        {
            await foreach (var resourceGroup in _resourceGroups.GetAllAsync())
            {
                var resourceGroupProto = new ResourceGroup
                {
                    Id = resourceGroup.Id.ToString(),
                    Name = resourceGroup.Data.Name ?? string.Empty,
                    Location = resourceGroup.Data.Location
                };
                resourceGroupList.Add(resourceGroupProto);
            }

            Log.Logger.WithCtx(context).Information("Resource groups found: {Count}", resourceGroupList.Count);
        }
        catch (Exception ex)
        {
            var grpcError = Server.HandleError(ex, "ListResourceGroups");
            Log.Logger.WithCtx(context).Error(grpcError, "Error occurred during resource group listing");
            throw grpcError;
        }

        return new ListResourceGroupResponse
        {
            RgList = { resourceGroupList }
        };
    }
}

File: mygreeterv3/server/internal/server/ListResourceGroups.go
package server

import (
	"context"
	"strconv"

	pb "dev.azure.com/service-hub-flg/service_hub_validation/_git/service_hub_validation_service.git/mygreeterv3/api/v1"
	"github.com/Azure/aks-middleware/grpc/server/ctxlogger"

	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armresources"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

func (s *Server) ListResourceGroups(ctx context.Context, in *pb.ListResourceGroupsRequest) (*pb.ListResourceGroupResponse, error) {
	logger := ctxlogger.GetLogger(ctx)
	if s.ResourceGroupClient == nil {
		logger.Error("ResourceGroupClient is nil in ListResourceGroups(), azuresdk feature is likely disabled")
		return nil, status.Errorf(codes.Unimplemented, "ResourceGroupClient is nil in ListResourceGroups(), azuresdk feature is likely disabled")
	}
	// creating a pager that is used to iterate over collection of resourceGroups'
	// Pass in nil to options parameter of NewListPager to get default pager
	pager := s.ResourceGroupClient.NewListPager(nil)
	var resourceGroups []*armresources.ResourceGroup
	var resourceGroupList []*pb.ResourceGroup
	for pager.More() {
		resp, err := pager.NextPage(ctx)
		if err != nil {
			logger.Error("NextPage() error: " + err.Error())
			return nil, HandleError(err, "NextPage()")
		}
		if resp.Value != nil {
			resourceGroups = append(resourceGroups, resp.Value...)
		}
	}

	logger.Info("Resource groups found: " + strconv.Itoa(len(resourceGroups)))

	for _, rg := range resourceGroups {
		resourceGroup := &pb.ResourceGroup{
			Id:       *rg.ID,
			Name:     *rg.Name,
			Location: *rg.Location,
		}
		resourceGroupList = append(resourceGroupList, resourceGroup)
	}
	return &pb.ListResourceGroupResponse{RgList: resourceGroupList}, nil
}
